json_serializable/lib/src/utils.dart

```diff
@@ -15,43 +15,61 @@ import 'package:source_gen/source_gen.dart';
 /// Dart code.
 // TODO: still need handle triple singe/double quotes within `value`
 String escapeDartString(String value) {
-  value = value.replaceAll('\\', r'\\');
+  var hasSingleQuote = false;
+  var hasDoubleQuote = false;
+  var hasDollar = false;
+  var canBeRaw = true;
+
   value = value.replaceAllMapped(_escapeRegExp, (match) {
+    var value = match[0];
+    if (value == "'") {
+      hasSingleQuote = true;
+      return value;
+    } else if (value == '"') {
+      hasDoubleQuote = true;
+      return value;
+    } else if (value == r'$') {
+      hasDollar = true;
+      return value;
+    }
+
+    canBeRaw = false;
     var mapped = _escapeMap[match[0]];
     if (mapped != null) return mapped;
     return _getHexLiteral(match[0]);
```

Comment:
Those three lines could be just:
```dart
return _escapeMap[value] ?? _getHexLiteral(value);
```
I'd even do:
```dart
return _escapeMap.putIfAbsent(value, _getHexLiteral);
```
(and make the map not constant, then you will cache the hex literals the first time you compute them instead of creating a new string each time).

Reply:
Ooo...not a bad idea...